### PR TITLE
Fixes EsriClient to work better with Seattle's find address candidates endpoint

### PR DESCRIPTION
**Pull request by seanmalbert**
### Description

Thre's an issue with Seattle's endpoint https://gisrevprxy.seattle.gov/arcgis/rest/services/locators/COS/GeocodeServer/findAddressCandidates does not work with the address parameters like the public endpoint does, despite the documentation saying it does https://gisdata.seattle.gov/server/sdk/rest/index.html#/Find_Address_Candidates/02ss00000015000000/.

So https://gisrevprxy.seattle.gov/arcgis/rest/services/locators/COS/GeocodeServer/findAddressCandidates?outFields=Address, SubAddr, City, RegionAbbr, Postal&f=json&address=609 Westlake Ave N&city=Seattle&region=WA&postal=85716 does not return what https://geocode.arcgis.com/arcgis/rest/services/World/GeocodeServer/findAddressCandidates?outFields=Address, SubAddr, City, RegionAbbr, Postal&f=json&address=609 Westlake Ave N&city=Seattle&region=WA&postal=85716 does.

This updates EsriClient to work with what the endpoint does return (which the standard public endpoint also does).

#### General

- [ ] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release >
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

#### User visible changes

- [ ] Wrote browser tests using the [validateAccessibility](https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/index.ts?L437:14&subtree=true) method
- [ ] Manually tested at 200% size
- [ ] Manually evaluated tab order

#### New Features

- [ ] Add new FeatureFlag env vars to `server/conf/application.conf`
- [ ] Conditioned new functionality on a [FeatureFlag](https://docs.civiform.us/contributor-guide/developer-guide/feature-flags)
- [ ] Wrote browser tests with the feature flag off and on, etc.

### Issue(s) this completes

Fixes #<issue_number>; Fixes #<issue_number>...

### CODE REVIEW

File: server/app/services/geo/esri/EsriClient.java [hunk 1]
```diff
@@ -144,11 +144,28 @@ CompletionStage<Optional<JsonNode>> fetchAddressSuggestions(ObjectNode addressJs
         Optional.ofNullable(addressJson.findPath(AddressField.STATE.getValue()).textValue());
     Optional<String> postal =
         Optional.ofNullable(addressJson.findPath(AddressField.ZIP.getValue()).textValue());
-    address.ifPresent(val -> request.addQueryParameter("address", val));
-    address2.ifPresent(val -> request.addQueryParameter("address2", val));
-    city.ifPresent(val -> request.addQueryParameter("city", val));
-    region.ifPresent(val -> request.addQueryParameter("region", val));
-    postal.ifPresent(val -> request.addQueryParameter("postal", val));
+    String singleLineAddress = "";
```
Comment (gwendolyngoetz) [inline server/app/services/geo/esri/EsriClient.java:147]: nonblocking: consider switching to StringBuilder

File: server/app/services/geo/esri/EsriClient.java [hunk 2]
```diff
@@ -144,11 +144,28 @@ CompletionStage<Optional<JsonNode>> fetchAddressSuggestions(ObjectNode addressJs
...
+      singleLineAddress += " " + city.get();
+    }
+
+    if (region.isPresent()) {
+      singleLineAddress += " " + region.get();
+    }
+
+    if (postal.isPresent()) {
+      singleLineAddress += " " + postal.get();
+    }
```
Comment (gwendolyngoetz) [inline server/app/services/geo/esri/EsriClient.java:166]: Does the address need to be formatted with commas to separate city, state, zip?

File: server/app/services/geo/esri/EsriClient.java [hunk 3]
```diff
@@ -330,7 +349,7 @@ public CompletionStage<ImmutableList<ServiceAreaInclusion>> getServiceAreaInclus
               List<String> features =
                   ctx.read(
                       "features[*].attributes." + esriServiceAreaValidationOption.getAttribute());
-
+              System.out.println("features = " + features);
```
Comment (gwendolyngoetz) [inline server/app/services/geo/esri/EsriClient.java]: Was this for debugging? Probably should remove it and use the logger if we need to keep the message.

File: server/app/services/geo/esri/EsriClient.java [hunk 4]
```diff
@@ -192,19 +209,21 @@ public CompletionStage<Optional<AddressSuggestionGroup>> getAddressSuggestions(A
...
-                        .setStreet(attributes.get("Address").asText())
-                        .setLine2(attributes.get("SubAddr").asText())
-                        .setCity(attributes.get("City").asText())
-                        .setState(attributes.get("RegionAbbr").asText())
-                        .setZip(attributes.get("Postal").asText())
+                        .setStreet(candidateJson.get("address").asText())
+                        .setLine2(attributes.get("SubAddr") == null ? address.getLine2() : attributes.get("SubAddr").asText())
+                        .setCity(attributes.get("City") == null || attributes.get("City").isEmpty() ? address.getCity() : attributes.get("City").asText())
+                        .setState(attributes.get("RegionAbbr") == null ? address.getState() : attributes.get("RegionAbbr").asText())
+                        .setZip(attributes.get("Postal") == null ? address.getZip() : attributes.get("Postal").asText())
```
Comment (gwendolyngoetz) [inline server/app/services/geo/esri/EsriClient.java:226]: Should all these null checks also have isEmpty checks? I recommend using Guava's com.google.common.base.Strings.isNullOrEmpty.  https://guava.dev/releases/21.0/api/docs/com/google/common/base/Strings.html#isNullOrEmpty-java.lang.String-

File: server/app/services/geo/esri/EsriClient.java (hunk 1 above)
Comment (seanmalbert) [reply server/app/services/geo/esri/EsriClient.java:147]: Good call. Updated!

File: server/app/services/geo/esri/EsriClient.java (hunk 2 above)
Comment (seanmalbert) [reply server/app/services/geo/esri/EsriClient.java]: It doesn't seem to matter, but I added a comma around city.

File: server/app/services/geo/esri/EsriClient.java (hunk 4 above)
Comment (seanmalbert) [reply server/app/services/geo/esri/EsriClient.java]: Unfortunately we can't cast a JsonNode to a string to use that without doing the null check first. I did add the empty checks to the other items as well.

**Review by gwendolyngoetz — APPROVED** [review]

**Comment by gwendolyngoetz** [comment]
Obsolete

**Pull request closed without merging**